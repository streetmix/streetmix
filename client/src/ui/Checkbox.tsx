/**
 * Custom stylized checkbox component, so we control the look and
 * feel instead of relying on browser's default styles.
 */
import React, { useState, useId, type ChangeEvent } from 'react'

import Icon from './Icon'
import './Checkbox.css'

interface CheckboxProps {
  // Child nodes are wrapped in <label> when rendered.
  children: React.ReactNode

  // Initial checked state of the input.
  checked?: boolean

  // Whether or not the input is disabled. Unlike vanilla HTML, this component
  // also changes the appearance of labels of disabled checkboxes.
  disabled?: boolean

  // The value of a checkbox input can be optionally set.
  // By default, browsers set checked inputs to have a string value of "on".
  value?: string

  // The handler function that's called when the value of the input changes.
  onChange?: React.ChangeEventHandler<HTMLInputElement>

  // Class name applied to containing element
  className?: string
}

function Checkbox (props: CheckboxProps): React.ReactElement {
  const {
    children,
    checked = false,
    disabled = false,
    value,
    onChange = () => undefined,
    className = '',

    // Remainder of props will be applied to the containing <div> element
    // for instance, `style`, data attributes, etc.
    ...restProps
  } = props
  // This is a controlled component. The `useState` hook maintains
  // this component's internal state, and sets the initial state
  // based on the `checked` prop (which is `false` by default).
  const [isChecked, setChecked] = useState(checked)

  // Generate an `id` to associate a `label` with an `input` element.
  const elementId = useId()

  const classNames = ['checkbox-item']
  if (className) {
    classNames.push(className)
  }

  // When the value is changed, we update the state, and we also call
  // any `onChange` handler that is provided by the parent via props.
  const handleChange = (event: ChangeEvent<HTMLInputElement>): void => {
    setChecked(!isChecked)
    onChange(event)
  }

  return (
    <div className={classNames.join(' ')} {...restProps}>
      <input
        type="checkbox"
        id={elementId}
        checked={isChecked}
        value={value}
        disabled={disabled}
        onChange={handleChange}
      />
      <label htmlFor={elementId}>{children}</label>
      {/* The visual state of this checkbox is affected by the value of the input, via CSS. */}
      <Icon name="check" />
    </div>
  )
}

export default Checkbox
