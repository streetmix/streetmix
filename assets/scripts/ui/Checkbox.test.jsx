import React from 'react'
import { vi } from 'vitest'
import { render } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import Checkbox from './Checkbox'

describe('Checkbox', () => {
  it('renders default snapshot', () => {
    const { asFragment } = render(<Checkbox>foo</Checkbox>)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders snapshot with all props', () => {
    const { asFragment } = render(
      <Checkbox
        checked={true}
        disabled={true}
        value="bar"
        id="baz"
        onChange={vi.fn()}
        className="custom-classname"
        style={{ color: 'blue' }}
      >
        foo
      </Checkbox>
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('handles click on the label text', async () => {
    const handleChange = vi.fn()
    const { getByText } = render(
      <Checkbox onChange={handleChange}>foo</Checkbox>
    )
    await userEvent.click(getByText('foo'))
    expect(handleChange).toHaveBeenCalledTimes(1)
  })
})
