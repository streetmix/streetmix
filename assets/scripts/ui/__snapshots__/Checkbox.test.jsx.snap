// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders default snapshot 1`] = `
<DocumentFragment>
  <div
    class="checkbox-item"
  >
    <input
      id="checkbox-id-1"
      type="checkbox"
      value=""
    />
    <label
      for="checkbox-id-1"
    >
      foo
    </label>
    <svg
      class="svg-inline--fa fa-check"
    />
  </div>
</DocumentFragment>
`;

exports[`Checkbox renders snapshot with all props 1`] = `
<DocumentFragment>
  <div
    class="checkbox-item custom-classname"
    style="color: blue;"
  >
    <input
      checked=""
      disabled=""
      id="baz"
      type="checkbox"
      value="bar"
    />
    <label
      for="baz"
    >
      foo
    </label>
    <svg
      class="svg-inline--fa fa-check"
    />
  </div>
</DocumentFragment>
`;
